samza-azure/src/main/java/org/apache/samza/system/eventhub/producer/EventHubSystemProducer.java

```diff
@@ -0,0 +1,337 @@
+    LOG.info("Trying to register {}.", streamName);
+    if (isStarted) {
+      String msg = "Cannot register once the producer is started.";
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    EventHubClientWrapper ehClient = eventHubClientWrapperFactory
+            .getEventHubClientWrapper(config.getStreamNamespace(streamName), config.getStreamEntityPath(streamName),
+                    config.getStreamSasKeyName(streamName), config.getStreamSasToken(streamName), config);
+
+    ehClient.init();
+    eventHubClients.put(streamName, ehClient);
+    streamPartitionSenders.put(streamName, new HashMap<>());
+  }
+
+  @Override
+  public synchronized void send(String destination, OutgoingMessageEnvelope envelope) {
+    if (!isStarted) {
+      throw new SamzaException("Trying to call send before the producer is started.");
+    }
+
+    if (!eventHubClients.containsKey(destination)) {
+      String msg = String.format("Trying to send event to a destination {%s} that is not registered.", destination);
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    if (sendExceptionOnCallback != null) {
+      SamzaException e = new SamzaException(sendExceptionOnCallback);
+      sendExceptionOnCallback = null;
+      pendingFutures.clear();
+      LOG.error("One of the previous sends failed.");
+      throw e;
+    }
+
+    EventData eventData = createEventData(destination, envelope);
+    int eventDataLength =  eventData.getBytes() == null ? 0 : eventData.getBytes().length;
+    eventWriteRate.get(destination).inc();
+    aggEventWriteRate.inc();
+    eventByteWriteRate.get(destination).inc(eventDataLength);
+    aggEventByteWriteRate.inc(eventDataLength);
+    EventHubClientWrapper ehClient = eventHubClients.get(destination);
+
+    Instant startTime = Instant.now();
+
+    CompletableFuture<Void> sendResult = sendToEventHub(destination, eventData, getEnvelopePartitionId(envelope),
+            ehClient.getEventHubClient());
+
+    Instant endTime = Instant.now();
+    long latencyMs = Duration.between(startTime, endTime).toMillis();
+    sendLatency.get(destination).update(latencyMs);
+    aggSendLatency.update(latencyMs);
+
+    long messageId = ++this.messageId;
+
+    // Rotate the messageIds
+    if (messageId == Long.MAX_VALUE) {
+      this.messageId = 0;
+    }
+
+    pendingFutures.put(messageId, sendResult);
+
+    // Auto remove the future from the list when they are complete.
+    sendResult.handle((aVoid, throwable) -> {
+        long callbackLatencyMs = Duration.between(endTime, Instant.now()).toMillis();
+        sendCallbackLatency.get(destination).update(callbackLatencyMs);
+        aggSendCallbackLatency.update(callbackLatencyMs);
+        if (throwable != null) {
+          sendErrors.get(destination).inc();
+          aggSendErrors.inc();
+          LOG.error("Send message to event hub: {} failed with exception: ", destination, throwable);
+          sendExceptionOnCallback = throwable;
```

Comment:
If this handler isn't called on the send thread (I assume it's called on the thread that completed the future?), this exception may still not be visible to the sender threads even if the send method is synchronized:

https://www.cs.umd.edu/~pugh/java/memoryModel/jsr-133-faq.html#synchronization
"Synchronization ensures that memory writes by a thread before or during a synchronized block are made visible in a predictable manner to other threads which synchronize <b>on the same monitor</b>."

Reply:
Good catch, I will to sync on the `sendExceptionOnCallback` throwable instead.